src/main/java/sparksoniq/jsoniq/runtime/iterator/operational/TreatIterator.java

```diff
@@ -35,43 +69,57 @@ public Item next() {
     }
 
     private void setNextResult() {
-        ItemType itemType = _sequenceType.getItemType();
-        String sequenceTypeName = ItemTypes.getItemTypeName(itemType.getType().toString());
-
         _nextResult = null;
-        if (_child.hasNext()) {
-            _nextResult = _child.next();
+        if (_iterator.hasNext()) {
+            _nextResult = _iterator.next();
+            if (_nextResult != null) _childIndex++;
         } else {
-            _child.close();
-            if (_childIndex == 0 && (_sequenceType.getArity() == SequenceType.Arity.One ||
-                    _sequenceType.getArity() == SequenceType.Arity.OneOrMore)) {
-                throw new TreatException(" Empty sequence cannot be treated as type " + sequenceTypeName + _sequenceType.getArity().getSymbol(), getMetadata());
-            }
+            _iterator.close();
+            checkEmptySequence(_childIndex);
         }
 
-        //... treat as ()
-        if (_nextResult != null && _sequenceType.isEmptySequence())
-            throw new TreatException(" " + ItemTypes.getItemTypeName(_nextResult.getClass().getSimpleName()) + " cannot be treated as type empty-sequence()", getMetadata());
+        this._hasNext = _nextResult != null;
+        if (!hasNext()) return;
 
-        //More items
-        if (_childIndex > 1 && (_sequenceType.getArity() == SequenceType.Arity.One ||
-                _sequenceType.getArity() == SequenceType.Arity.OneOrZero)) {
-            throw new TreatException(" Sequences of more than one item cannot be treated as type " + sequenceTypeName + _sequenceType.getArity().getSymbol(), getMetadata());
+        checkItemsSize(_childIndex);
+        if (!_nextResult.isTypeOf(itemType)) {
+            throw new TreatException(" " + ItemTypes.getItemTypeName(_nextResult.getClass().getSimpleName()) + " cannot be treated as type " + sequenceTypeName + _sequenceType.getArity().getSymbol(), getMetadata());
         }
+    }
 
-        if (_nextResult != null && !_nextResult.isTypeOf(itemType)) {
-            throw new TreatException(" " + ItemTypes.getItemTypeName(_nextResult.getClass().getSimpleName()) + " cannot be treated as type " + sequenceTypeName + _sequenceType.getArity().getSymbol(), getMetadata());
+    @Override
+    public JavaRDD<Item> getRDD(DynamicContext dynamicContext) {
+        _currentDynamicContext = dynamicContext;
+        JavaRDD<Item> childRDD = _iterator.getRDD(dynamicContext);
+
+        long count = childRDD.count();
```

Comment:
There is optimization potential here: you could use an action that only takes up to two items, and use the local count of what is returned (0, 1 or 2). Two indicates more than one in the original sequence, and there is no need for the exact count for the purpose of this iterator.